File: app/src/main/java/com/offlinetranslator/app/feature/learn/SrsScheduler.kt
package com.offlinetranslator.app.feature.learn

/** SRS 间隔重复纯函数调度器（Leitner 盒子 + 二选一）。无 Android 依赖，纯单测。 */
object SrsScheduler {
    val intervalsDays = longArrayOf(0, 1, 3, 7, 16, 30)
    const val MAX_BOX = 6
    const val DAY_MS = 86_400_000L
    const val FAR_FUTURE = Long.MAX_VALUE

    data class Update(val box: Int, val dueAt: Long, val missCount: Int, val lastReviewedAt: Long)

    /** 给一次复习评分，算出新的调度状态。correct=认识 / !correct=不认识。 */
    fun schedule(box: Int, missCount: Int, correct: Boolean, now: Long): Update {
        if (!correct) {
            return Update(box = 0, dueAt = now, missCount = missCount + 1, lastReviewedAt = now)
        }
        val newBox = (box + 1).coerceAtMost(MAX_BOX)
        val dueAt = if (newBox >= MAX_BOX) FAR_FUTURE
        else now + intervalsDays[newBox] * DAY_MS
        return Update(box = newBox, dueAt = dueAt, missCount = missCount, lastReviewedAt = now)
    }
}

/** 连续打卡纯逻辑。用「本地日序号」(epochDay) 计算，避免时区耦合。 */
object StreakLogic {
    data class State(val lastStudyDay: Long, val currentStreak: Int)

    /** 完成≥1张复习时调用。today/lastDay 为本地日序号。 */
    fun onStudied(lastDay: Long, current: Int, today: Long): State = when {
        lastDay == today -> State(today, current)          // 今天已记过
        lastDay == today - 1 -> State(today, current + 1)  // 昨天→续
        else -> State(today, 1)                            // 断档/首次→重置
    }

    /** 展示用：若 lastDay 非今天也非昨天，连续已断，显示 0。 */
    fun displayStreak(lastDay: Long, current: Int, today: Long): Int =
        if (lastDay == today || lastDay == today - 1) current else 0
}

/** 今日到期池筛选纯逻辑。 */
object DuePool {
    /**
     * 从全部卡里选出今日该练的：
     *  - 到期卡：dueAt<=now 且 (box>=1 或 lastReviewedAt>0)  含答错打回但没练完的
     *  - 新卡：box==0 且 lastReviewedAt==0，按 newLimit 取前 N（按传入顺序）
     */
    fun <T> select(
        cards: List<T>, now: Long, newLimit: Int,
        box: (T) -> Int, dueAt: (T) -> Long, lastReviewedAt: (T) -> Long,
    ): List<T> {
        val due = cards.filter { dueAt(it) <= now && (box(it) >= 1 || lastReviewedAt(it) > 0) }
        val fresh = cards.filter { box(it) == 0 && lastReviewedAt(it) == 0L }.take(newLimit)
        return due + fresh
    }
}

File: ios/Yiren/Features/Learn/SrsScheduler.swift
import Foundation

/// SRS 间隔重复纯函数调度器（与 Android SrsScheduler.kt 逐字一致）。
enum SrsScheduler {
    static let intervalsDays: [Int64] = [0, 1, 3, 7, 16, 30]
    static let maxBox = 6
    static let dayMs: Int64 = 86_400_000

    struct Update { let box: Int; let dueAt: Int64; let missCount: Int; let lastReviewedAt: Int64 }

    static func schedule(box: Int, missCount: Int, correct: Bool, now: Int64) -> Update {
        if !correct { return Update(box: 0, dueAt: now, missCount: missCount + 1, lastReviewedAt: now) }
        let newBox = min(box + 1, maxBox)
        let dueAt = newBox >= maxBox ? Int64.max : now + intervalsDays[newBox] * dayMs
        return Update(box: newBox, dueAt: dueAt, missCount: missCount, lastReviewedAt: now)
    }
}

enum StreakLogic {
    struct State { let lastStudyDay: Int64; let currentStreak: Int }
    static func onStudied(lastDay: Int64, current: Int, today: Int64) -> State {
        if lastDay == today { return State(lastStudyDay: today, currentStreak: current) }
        if lastDay == today - 1 { return State(lastStudyDay: today, currentStreak: current + 1) }
        return State(lastStudyDay: today, currentStreak: 1)
    }
    static func displayStreak(lastDay: Int64, current: Int, today: Int64) -> Int {
        (lastDay == today || lastDay == today - 1) ? current : 0
    }
}

enum DuePool {
    static func select<T>(cards: [T], now: Int64, newLimit: Int,
                          box: (T) -> Int, dueAt: (T) -> Int64, lastReviewedAt: (T) -> Int64) -> [T] {
        let due = cards.filter { dueAt($0) <= now && (box($0) >= 1 || lastReviewedAt($0) > 0) }
        let fresh = cards.filter { box($0) == 0 && lastReviewedAt($0) == 0 }.prefix(newLimit)
        return due + Array(fresh)
    }
}
